Make row color alternating in the Peers tab optional

hebasto opened the pull request:
This PR:
- makes an "alternating row colors" feature in the Peers tab switchable to satisfy all users with different tastes.
- is a follow up of #298
- is an alternative to #306

Screenshot:

![Screenshot from 2021-05-01 18-56-44](https://user-images.githubusercontent.com/32963518/116787820-0718da00-aaaf-11eb-909d-26caa2f09cfc.png)

hebasto commented:
cc: @rebroad @jonatack @RandyMcMillan @Bosch-0 @jarolrod

jonatack commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -805,6 +805,16 @@
+       <item>
+        <widget class="QCheckBox" name="peersTabAlternatingRowColors">
+         <property name="toolTip">
+          <string>Turn on/off row color alternating for both &quot;Peers&quot; and &quot;Banned peers&quot; tables in the Peer Tab.</string>
```
```suggestion
          <string>Alternate the row colors for the &quot;Peers&quot; and &quot;Banned peers&quot; tables in the Peers tab.</string>
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/307#issuecomment-830652956).

jonatack reviewed (commented):
Concept ACK, thank you for adding this flexibility.

hebasto commented:
Updated 8551215a071244ed28fa4b2f426a1dab82743bf8 -> fe7d615cd8d746d6b36e3243f10ae5b954d1e544 ([pr307.01](https://github.com/hebasto/gui/commits/pr307.01) -> [pr307.02](https://github.com/hebasto/gui/commits/pr307.02), [diff](https://github.com/hebasto/gui/compare/pr307.01..pr307.02)):

- addressed @jonatack's [comment](https://github.com/bitcoin-core/gui/pull/307#discussion_r624527158)

jonatack commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -805,6 +805,16 @@
+          <string>Alternate the row colors for the &quot;Peers&quot; and &quot;Banned peers&quot; tables in the Peers tab.</string>
+         </property>
+         <property name="text">
+          <string>Alternate row colors in the Peers Tab</string>
```
s/Tab/tab/

jonatack replied:
(or change Overview s/tab/Tab/ but lowercase seems fine)

hebasto replied:
Thanks. [Updated](https://github.com/bitcoin-core/gui/pull/307#issuecomment-830653215).

hebasto commented:
Updated fe7d615cd8d746d6b36e3243f10ae5b954d1e544 -> b7de85d51c503576514d6d9cef0fbc2eb00e8aea ([pr307.02](https://github.com/hebasto/gui/commits/pr307.02) -> [pr307.03](https://github.com/hebasto/gui/commits/pr307.03), [diff](https://github.com/hebasto/gui/compare/pr307.02..pr307.03)):

- addressed @jonatack's [comment](https://github.com/bitcoin-core/gui/pull/307#discussion_r624527938)

rebroad commented:
I'm sorry for being picky..! Or a diva... or whatever it is...

hebasto commented:
>  1. No stripes by default, but a config option somewhere more hidden than this proposed option (maybe in a config file instead?)

1. What is the rationale of hiding a visual appearance option from users?

2. The `bitcoin.conf` is not dedicated for such options.

promag commented:
So far we had alternate rows and nobody complained AFAIK. If we add this option I think it should affect all tables, not only peers table.

Concept ACK figuring out the scope of appearance settings. Until then NACK, this particular option seems to have no grounds and might lead to similar pull requests.

jonatack commented:
I think this is a fantastic first step back to a cleaner UI and it would be even better as a global option in a follow-up.

jonatack commented:
(I have always avoided using the tables that had stripes before. They play havoc with the useability for me; my eyesight doesn't like it. Fortunately I can get by fine with the CLI if people like striping the GUI background all over, but it might be good to consider questions of ableism and accessibility in the UI.)

jarolrod commented:
I agree with this 

> If we add this option I think it should affect all tables, not only peers table.

Also, this sets a good precedent for continuing to make the GUI user configurable.

RandyMcMillan commented:
Concept ACK

hebasto commented:
Rebased b7de85d51c503576514d6d9cef0fbc2eb00e8aea -> 081842d6a735a85e5611ab36703b63f8632f7f4b ([pr307.03](https://github.com/hebasto/gui/commits/pr307.03) -> [pr307.04](https://github.com/hebasto/gui/commits/pr307.04)) due to the conflict with #194.

jarolrod reviewed (commented):
ACK 081842d6a735a85e5611ab36703b63f8632f7f4b

Thanks for adding this option and tidying up the spacing in `optionsdialog.ui`


> If we add this option I think it should affect all tables, not only peers table.

This can be addressed in a follow-up PR.

**Start up GUI**: 

| Start           | Enable Option |
| ----------- | -------------- |
| ![Screen Shot 2021-05-12 at 11 12 33 PM](https://user-images.githubusercontent.com/23396902/118072622-39acb780-b378-11eb-9ae3-468e0a8f06c1.png) | ![Screen Shot 2021-05-12 at 11 13 15 PM](https://user-images.githubusercontent.com/23396902/118072658-48936a00-b378-11eb-9efd-58ce58b28cb0.png) |

**Restart GUI:**

| Start: Setting Persists ✅       | Disable Option |
| ----------------------------- | --------------- |
| ![Screen Shot 2021-05-12 at 11 20 19 PM](https://user-images.githubusercontent.com/23396902/118072907-ce171a00-b378-11eb-81b1-3fd2dc575fe9.png) | ![Screen Shot 2021-05-12 at 11 20 37 PM](https://user-images.githubusercontent.com/23396902/118072916-d2433780-b378-11eb-8b9f-b911b888754c.png) |

hebasto commented:
Rebased 081842d6a735a85e5611ab36703b63f8632f7f4b -> 2bc53b62494b24388923aa2697443949b8bbbba7 ([pr307.04](https://github.com/hebasto/gui/commits/pr307.04) -> [pr307.05](https://github.com/hebasto/gui/commits/pr307.05)) due to the conflict with #123.

hebasto commented:
Rebased 2bc53b62494b24388923aa2697443949b8bbbba7 -> 3231689fa9b379e2c214a69b6694340f22c2a7ae ([pr307.05](https://github.com/hebasto/gui/commits/pr307.05) -> [pr307.06](https://github.com/hebasto/gui/commits/pr307.06)) due to the conflict with #325.

jonatack reviewed (commented):
Tested ACK 3231689fa9b379e2c214a69b6694340f22c2a7ae modulo that it would be nice for the change to occur immediately when checking/unchecking the option and I would squash the last 2 commits.

Wishlist:
- The options dialog opens to the last tab used
- Further options for toggling alternate row coloring in other places (or globally)

hebasto commented:
Rebased 3231689fa9b379e2c214a69b6694340f22c2a7ae -> 84be4844064d8e1bc973a5f5c662d2069589e1c2 ([pr307.06](https://github.com/hebasto/gui/commits/pr307.06) -> [pr307.07](https://github.com/hebasto/gui/commits/pr307.07)) due to the conflict with #256.

jarolrod reviewed (commented):
ACK 84be484

Tested on `Arch Linux Qt 5.15.2`, `macOS 11.3 Qt 5.15.2`, and cross-compile to `Windows 10`.

**Notes on commits:**
- [x] ada34cfee3f8696d85f75353a0979effe194746b
  -  Confirmed that the new indentation is correct
- [x] 217d42b7465bf210511ef6cdff0475eb0a9e078c
  -  Code Review ACK, couple of nits ([1](https://github.com/bitcoin-core/gui/pull/307#discussion_r646046898), [2](https://github.com/bitcoin-core/gui/pull/307#discussion_r646047100))
- [x] 84be4844064d8e1bc973a5f5c662d2069589e1c2
  -  Cherry-picking the first two commits then running `clang-format-diff.py` gives me the same diff      

**Screenshots:**
| Start      | Enable  |
| ----------- | ----------- |
| ![307-start](https://user-images.githubusercontent.com/23396902/120907643-3633e580-c631-11eb-80e9-669c18f9254a.png) | ![307-enable](https://user-images.githubusercontent.com/23396902/120907648-39c76c80-c631-11eb-8451-c5d29134562f.png) |

| Restart  | Disable  |
| ----------- | ------------ |
| ![307-restart](https://user-images.githubusercontent.com/23396902/120907664-582d6800-c631-11eb-9f34-0a92dcfa1932.png) | ![307-disable](https://user-images.githubusercontent.com/23396902/120907671-5cf21c00-c631-11eb-90a6-fc2da774af3a.png) |

jarolrod commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -6,8 +6,8 @@
     <y>0</y>
-    <width>560</width>
-    <height>440</height>
+    <width>646</width>
```
Nit

 there's not really a reason to change the size here. I guess this is a good place to answer a question:

"When adding new settings, should we maintain a certain level of padding between the last setting option and its enclosing box?"

| Use PR Size | Use Master Size |
| ----------------- | ----------------------- |
| ![size-increase](https://user-images.githubusercontent.com/23396902/120907709-bbb79580-c631-11eb-8755-6d7bac8f0f76.png) | ![size-normal](https://user-images.githubusercontent.com/23396902/120907710-bf4b1c80-c631-11eb-85b8-f4a7d2f3cca2.png) |

hebasto replied:
> there's not really a reason to change the size here. I guess this is a good place to answer a question:

The reason is that successive openings and closings of the `optionsdialog.ui` in Qt Designer won't suggest size changes.

jarolrod commented on src/qt/optionsmodel.h:
```diff
@@ -44,29 +44,30 @@ class OptionsModel : public QAbstractListModel
+        DatabaseCache,                // int
+        SpendZeroConfChange,          // bool
+        Listen,                       // bool
         OptionIDRowCount,
```
Nit in 217d42b7465bf210511ef6cdff0475eb0a9e078c

While here, you could document the type here is `int`.

Additionally, if you have to retouch, you could rename the commit from:
`gui: Add "Alternating Row Color" settings for the Peers Tab `

to:

`qt: Add "Alternating Row Color" settings for the Peers Tab `

hebasto replied:
> While here, you could document the type here is `int`.

I don't think it is required to document the type of a counter, as it is not an option id.

hebasto replied:
> Additionally, if you have to retouch, you could rename the commit from:
> `gui: Add "Alternating Row Color" settings for the Peers Tab `
> 
> to:
> 
> `qt: Add "Alternating Row Color" settings for the Peers Tab `

Done in the [recent](https://github.com/bitcoin-core/gui/pull/307#issuecomment-859324593) push.

hebasto commented:
Rebased 84be4844064d8e1bc973a5f5c662d2069589e1c2 -> fdf80937d1cc2eb7faddbc6d5ffcdb66659fb758 ([pr307.07](https://github.com/hebasto/gui/commits/pr307.07) -> [pr307.08](https://github.com/hebasto/gui/commits/pr307.08)) due to the conflict with #4.

wodry commented:
> So far we had alternate rows and nobody complained AFAIK. If we add this option I think it should affect all tables, not only peers table.
> 
> Until then NACK, this particular option seems to have no grounds and might lead to similar pull requests.

Also NACK. First, alternate rows are added, then an option is added to disable it. Do not start option bloating because of groundless "tastes" of some users. Keep it simple (and unflexible).

jonatack commented:
If we don't want to have options for accessibility then ISTM the striping should be removed globally.

Rspigler commented:
Sad to see this closed.  Concept ACK from me.
